contractcourt/contract_resolvers.go

```diff
@@ -585,6 +602,21 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
 		return nil, fmt.Errorf("quitting")
 	}
 
+	// Finally, with the HTLC claimed, we'll attempt to settle its
+	// corresponding invoice if we were the original destination. We'll
+	// retrieve the preimage from the success transaction's witness and
+	// settle the invoice with the output's amount, taking into account
+	// on-chain fees (if any).
+	paymentHash := sha256.Sum256(h.htlcResolution.Preimage[:])
+	amt := btcutil.Amount(h.htlcResolution.SignedSuccessTx.TxOut[0].Value)
```

Comment:
Here we take as amount the output value of the success tx, which is the paid amount minus the success tx fee. 

This amount will be reduced further by the nursery sweep which needs fees too, before the funds are back in the wallet. Nursery may sweep multiple inputs in a single tx, so difficult to attribute (part) of the fee to this invoice payment.

However, in the remote commitment case above, as invoice amount the paid value minus the sweep fee is recorded. Looks inconsistent? Also it will be a problem when that sweep is (soon) also combined with other inputs.

From the perspective of the sender of the payment, subtracting any on chain fees may not be justifiable. It may look like not the full amount has been paid, while it isn't his fault that the second-last hop caused channel closure.

Reply:
Agreed - fixed. Now we'll use just the value of the output we're sweeping directly. One thing I can see come up in the future is node operators wondering where the remaining amount went to when sweeping the HTLC on-chain. Currently (I think), there isn't a way to distinguish an invoice being settled off-chain vs on-chain.

Reply:
I think it is safer to settle the invoice with the amount on the commitment tx, excluding both stage 1 and stage 2 fees. To indeed prevent invoices from marked as settled in the db with a value less than the original invoice amt. But we should do that in both the local and remote close case.

Reply:
IMO it's correct to use the output amount directly, and not the original value of the output within the commitment. If we use the original HTLC value, then from an external accounting point of view, we create an inconsistent state. We'd show the HTLC as being settled with a value of 1k satoshis, but then by the time we actually sweep it, it may be 500 or so satoshis (just an example). We already have the `AmtPaid` field in the database and can use that along with a new settle state `ContractSettledChain` to indicate that we had to go on chain to sweep it. Once the resolvers are fully connected to the sweeper, then the amount swept is even hazier as part of the HTLC amount will go to fees for a batched sweep.

Reply:
> Once the resolvers are fully connected to the sweeper, then the amount swept is even hazier as part of the HTLC amount will go to fees for a batched sweep.

In this case, how would we go about determining the settled amount (HTLC amount - chain fees)? We could weigh all the inputs and determine the settled amount with a ratio, but not sure if that's worth the hassle.

Reply:
Would the `AmtPaid` field contain the original HTLC value then? I am worried that if we don't do that, it could be that the sender paid his invoice in full, but the final amount was lower because the receiver needed to go on chain. This is to be blamed on an intermediate node, but it may look like the sender didn't fulfill his obligation.

Reply:
That's why we would add a new settle state `ContractSettledChain` to signal this to the receiver. We can also make it clear in the docs that when an invoice has been settled with this state, the amount received will be less due to chain fees.

Reply:
Yeah to circle back on this we'd add a new settle state, as generally accurate fee accounting in the face of batching is a whole project in itself.

Reply:
I now realize that once we do hold invoice, we'll also need to implement the accepted state in the contract resolvers. That'll be another settled state then, `ContractAcceptedChain`.

Reply:
Another way to look at it, could be similar to traditional banking: I pay various service fees to my bank and I book them as a separate cost. I am not trying to attribute those to the individual invoices I sent out. Could we do it similarly for lnd? Keep track of a global costs administration that includes chain fees and book invoices with the original htlc value?

Reply:
Circled back on this and we decided to just use the full HTLC value for now. We'll then defer some sort of fee audit functionality for a channel that can tell you the amount paid in on-chain fees from open to close of the channel to a follow up PR.

Reply:
Fixed. Realized we can't even use the HTLC value as we don't persist it on disk and there doesn't seem to be a way to retrieve it from what's currently stored (something we might want to store in the future perhaps?). As an alternative, I went with just settling the invoice with whatever it originally requested. I believe this is still fine to use as the value is inaccurate anyway.